File: apps/executor-python/manifest.py
"""manifest.yaml parsing utility"""
import logging
from pathlib import Path
from typing import Any

try:
    import yaml
except ImportError:
    yaml = None  # type: ignore

logger = logging.getLogger(__name__)


def load_manifest(work_dir: Path) -> dict[str, Any]:
    """Load manifest.yaml from the working directory. Returns an empty dict if not found."""
    manifest_path = work_dir / 'manifest.yaml'
    if not manifest_path.exists():
        manifest_path = work_dir / 'manifest.yml'
    if not manifest_path.exists():
        return {}

    if yaml is None:
        logger.warning('PyYAML not installed, cannot parse manifest.yaml')
        return {}

    try:
        with open(manifest_path, 'r', encoding='utf-8') as f:
            data = yaml.safe_load(f) or {}
        logger.info(f'Loaded manifest: {manifest_path}')
        return data
    except Exception as e:
        logger.warning(f'Failed to parse manifest: {e}')
        return {}


def merge_task_with_manifest(task: dict, manifest: dict) -> dict:
    """Fill in missing task fields with manifest values (task fields take priority)."""
    merged = {**manifest, **task}
    # Merge requirements (both manifest and task may have them)
    m_reqs = manifest.get('requirements', [])
    t_reqs = task.get('requirements', [])
    if m_reqs or t_reqs:
        merged['requirements'] = list(dict.fromkeys(m_reqs + t_reqs))
    return merged

File: apps/executor-node/src/manifest.ts
import * as fs from 'fs';
import * as path from 'path';
import * as yaml from 'js-yaml';
import { logger } from './logger';

export interface Manifest {
  runtime?: string;
  entrypoint?: string;
  timeout?: number;
  requirements?: string[];
  [key: string]: unknown;
}

/**
 * Load manifest.yaml / manifest.yml from the working directory.
 * Returns an empty object if the file does not exist or fails to parse.
 */
export function loadManifest(workDir: string): Manifest {
  for (const name of ['manifest.yaml', 'manifest.yml']) {
    const p = path.join(workDir, name);
    if (fs.existsSync(p)) {
      try {
        const content = fs.readFileSync(p, 'utf-8');
        const data = yaml.load(content) as Manifest;
        logger.info(`Loaded manifest: ${p}`);
        return data || {};
      } catch (e: any) {
        logger.warn(`Failed to parse manifest ${p}: ${e.message}`);
      }
    }
  }
  return {};
}

/**
 * Fill in missing task fields with manifest values (task fields take priority).
 * Merge and deduplicate the requirements list.
 */
export function mergeTaskWithManifest(
  task: Record<string, unknown>,
  manifest: Manifest,
): Record<string, unknown> {
  const merged: Record<string, unknown> = { ...manifest, ...task };
  const mReqs: string[] = (manifest.requirements as string[]) || [];
  const tReqs: string[] = (task.requirements as string[]) || [];
  if (mReqs.length || tReqs.length) {
    merged.requirements = [...new Set([...mReqs, ...tReqs])];
  }
  return merged;
}
